prov/net/src/xnet_progress.c

```diff
@@ -91,10 +108,94 @@ static int xnet_io_uring_fd(struct xnet_io_uring *io_uring)
 	assert(xnet_io_uring);
 	return io_uring->ring.ring_fd;
 }
+
+static bool xnet_io_uring_needs_submit(struct xnet_io_uring *io_uring)
+{
+	return io_uring_sq_ready(&io_uring->ring);
+}
+
+static void xnet_submit_io_uring(struct xnet_io_uring *io_uring)
+{
+	if (xnet_io_uring && io_uring_sq_ready(&io_uring->ring))
+		io_uring_submit(&io_uring->ring);
+}
+
+static bool xnet_get_io_uring_credit(struct xnet_io_uring *io_uring)
+{
+	if (!xnet_io_uring)
+		return true;
+
+	if (io_uring->credits > 0) {
+		io_uring->credits--;
+		return true;
+	}
+	return false;
+}
+
+static void xnet_progress_cqe(struct xnet_io_uring *io_uring,
+			      struct io_uring_cqe *cqe)
+{
+	struct xnet_xfer_entry *tx_entry;
+	struct ofi_bsock *bsock;
+	struct xnet_ep *ep;
+
+	assert(xnet_io_uring);
+	bsock = (struct ofi_bsock *) cqe->user_data;
+	assert(bsock);
+
+	if (&io_uring->ring == bsock->sockapi->tx_io_uring)
+	{
+		ep = container_of(bsock, struct xnet_ep, bsock);
+		tx_entry = ep->cur_tx.entry;
+		assert(tx_entry);
+		assert(ep->cur_tx.io_uring_busy);
+
+		ep->cur_tx.io_uring_busy = false;
+		if (cqe->res < 0) {
+			if (!OFI_SOCK_TRY_SND_RCV_AGAIN(-cqe->res))
+				xnet_complete_tx(ep, cqe->res);
+		} else {
+			ep->cur_tx.data_left -= cqe->res;
+			if (ep->cur_tx.data_left)
+				ofi_consume_iov(tx_entry->iov, &tx_entry->iov_cnt,
+						cqe->res);
+			else
+				xnet_complete_tx(ep, FI_SUCCESS);
+		}
+		xnet_progress_tx(ep);
+	}
+}
+
+static void xnet_progress_io_uring(struct xnet_io_uring *io_uring)
+{
+	struct io_uring_cqe *cqes[XNET_MAX_EVENTS];
+	int nready;
+	int i;
+
+	if (!xnet_io_uring)
+		return;
+
+	nready = io_uring_peek_batch_cqe(&io_uring->ring, cqes, XNET_MAX_EVENTS);
+	if (!nready)
+		return;
+
+	assert(nready <= XNET_MAX_EVENTS);
+	for (i = 0; i < nready; i++) {
+		xnet_progress_cqe(io_uring, cqes[i]);
+		io_uring->credits++;
```

Comment:
Is there a potential problem here with this ordering?  We have a completion for a send, but the send may not be done.  That falls through to xnet_progress_tx(), which may need to post another transfer, which in turns need another credit.  Could we temporarily block the send because we haven't updated the credits?

Reply:
Thanks for catching that! You are right and the credit must be released before executing `xnet_progress_cqe()`. I will update the code.